stats.c

```diff
@@ -2020,34 +2022,31 @@ int replicate_regions(stats_t *stats, hts_itr_multi_t *iter, stats_info_t *info)
     return 0;
 }
 
-void init_group_id(stats_t *stats, const char *id)
+static void init_group_id(stats_t *stats, stats_info_t *info, const char *id)
 {
-#if 0
-    if ( !stats->sam_header->dict )
-        stats->sam_header->dict = sam_header_parse2(stats->sam_header->text);
-    void *iter = stats->sam_header->dict;
-    const char *key, *val;
-    int n = 0;
-    stats->rg_hash = khash_str2int_init();
-    while ( (iter = sam_header2key_val(iter, "RG","ID","SM", &key, &val)) )
-    {
-        if ( !strcmp(id,key) || (val && !strcmp(id,val)) )
-        {
-            khiter_t k = kh_get(kh_rg, stats->rg_hash, key);
-            if ( k != kh_end(stats->rg_hash) )
-                fprintf(stderr, "[init_group_id] The group ID not unique: \"%s\"\n", key);
-            int ret;
-            k = kh_put(kh_rg, stats->rg_hash, key, &ret);
-            kh_value(stats->rg_hash, k) = val;
-            n++;
+    stats->rg_hash = kh_init(rg);
+    if (!stats->rg_hash) return;
+    sam_hdr_t *hdr = info->sam_header;
+    const char *key;
+    kstring_t sm = KS_INITIALIZE;
+    int i, ret, nrg = sam_hdr_count_lines(hdr, "RG");
+
+    for (i=0; i<nrg; i++) {
+        key = sam_hdr_line_name(hdr, "RG", i);
+        if (!strcmp(key, id)) {
+            kh_put(rg, stats->rg_hash, key, &ret);
+            if (ret == -1) { ks_free(&sm); return; }
+        } else { /* Check for SM name, as per manual */
+            if (!sam_hdr_find_tag_pos(hdr, "RG", i, "SM", &sm)) {
+                if (ks_str(&sm) && !strcmp(ks_str(&sm), id)) {
```

Comment:
An alternative is to use `ks_c_str(&sm)` to ensure you never get NULL.

Reply:
Thanks. I don't think the bug comes from there, though.